File: 0653-two-sum-iv-input-is-a-bst/0653-two-sum-iv-input-is-a-bst.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public boolean solveFindTarget(TreeNode root, int k, Set<Integer> set){
        if(root == null){
            return false;
        }
        if(set.contains(k - root.val)){
            return true;
        }
        set.add(root.val);
        return solveFindTarget(root.left, k, set) || solveFindTarget(root.right, k, set);
    }
    public boolean findTarget(TreeNode root, int k) {
        Set<Integer> set = new HashSet<>();
        return solveFindTarget(root, k, set);
    }
}

File: 0653-two-sum-iv-input-is-a-bst/0653-two-sum-iv-input-is-a-bst.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    bool solveFindTarget(TreeNode* root, int k, set<int> &set){
        if(root == nullptr){
            return false;
        }
        if(set.find(k - root->val) != set.end()){
            return true;
        }
        set.insert(root->val);
        return solveFindTarget(root->left, k, set) || solveFindTarget(root->right, k, set);
    }
    bool findTarget(TreeNode* root, int k) {
        set<int> set;
        return solveFindTarget(root, k, set);
    }
};